[3.x] Shader goodies: async. compilation + caching (ubershader approach)

RandomShaper opened the pull request:
This PR, by being a better implementation of #46330, supersedes it. All the feedback from maintainers that was sent there has been addressed in this one. The approach to a reasonable fallback during background compilation that this new implementation features is ubershaders (details below), as suggested by @reduz.

---

The main goal of this PR is to reduce stalling in games.

![image](https://user-images.githubusercontent.com/11797174/137089487-45a4fda6-14f2-42f1-9169-201be87273ca.png)
**UPDATE:** Screenshots updated to reflect latest changes.

# Shader caching

As long as the target platform supports the program binary GL extension, this is just enable and forget.

Some remarks:
- ~~Writes to the cache are async. to prevent stalling the render thread as much as possible.~~ **UPDATE:**  The cache is now only used for ubershaders (see below).
- If async. shader compilation is enabled in addition to caching, shader "reconstruction" from its cache file is also potentially done asynchronously.

# Asynchronous compilation of shaders

It will work if enabled and supported by the GL driver. If native parallel compilation is supported, that's used, which is the most efficient. Otherwise, asynchronicity is achieved via a secondary GL context (and another thread) that sends the compiled shader back to the main one in its binary form, which means the program binary extension must be supported. If both fail, async. compilation is effectively disabled.

~~Three fallback modes are added to both manually created shaders (either codey or visual) and `SpatialMaterial`s: _ubershader_, _none_ and _no render_.~~ **UPDATE:** Two fallback modes are added to both manually created shaders (either codey or visual) and `SpatialMaterial`s: _visible (ubershader)_, _hidden_.
Please check the diff where these are explained in the built-in documentation.

The default mode is _visible_, which is the main highlight of this PR. It hints the engine to use a universal version of the base shader (e.g., scene) that supports all the possible contexts (rendering conditionals), so it's slower than the properly conditioned version of the real shader, but ready from the very beginning to be used while the real deal is being compiled in the background.

Some simple parsing of special directives has been added to the GLSL language so the base shaders can be transformed on the go from a  preprocessor based handling of conditionals to a runtime one.

 You can explicitly set a different conservative mode for any shader/material.

**UPDATE:** Screenshots updated to reflect latest changes.

![image](https://user-images.githubusercontent.com/11797174/137089615-0e8bdfae-30f2-4fbd-956f-ecf4fb3753c5.png)
![image](https://user-images.githubusercontent.com/11797174/137089924-557382af-7f1c-4cb9-bace-d250501441a0.png)
![image](https://user-images.githubusercontent.com/11797174/137090523-7b946f12-aa87-4f4f-8a6f-e595b087386b.png)

Please also see the diff for an explanation of the different project settings.

![image](https://user-images.githubusercontent.com/11797174/137093987-fa037cfa-5c7e-4a94-a2a6-9fb9a81c1070.png)

# Testing

Since both caching and async. compilation are experimental at this point, they are disabled by default. Please test this in your own projects with and without caching. A good starting point would be to enable caching for the editor only and keep the default number of simultaneous compiles. Depending on the target hardware, among other things, those may be raised to find the perfect balance for any project.

# Future work

These would be good for future PRs:

- Ubershaderify other shaders than the scene one (~particles~ (DONE), canvas).
- ~Split vertex and shader compilation into discrete states, for finer grained control of CPU usage.~ (DONE)
- Collect/provide metrics about fallbacks used, compiles in flight, cache retrievals, etc.
- Allow the ubershader to be customized to control the balance between quality and performance.

RandomShaper committed:
Avoid the need for copy assignment in HashMap key/data types

RandomShaper committed:
Implement get_cache_path() for iOS, and improve it for Android and Windows

RandomShaper committed:
Add off-screen GL context

RandomShaper committed:
Implement async shader compilation plus caching for GL ES 3

Async. compilation via ubershader is currently available in the scene and particles shaders only.

Bonus:
- Use `#if defined()` syntax for not true conditionals, so they don't unnecessarily take a bit in the version flagset.
- Remove unused `ENABLE_CLIP_ALPHA` from scene shader.
- Remove unused `PARTICLES_COPY` from the particles shader.
- Remove unused uniform related code.
- Shader language/compiler: use ordered hash maps for deterministic code generation (needed for caching).

Calinou commented on doc/classes/SpatialMaterial.xml:
```diff
@@ -173,6 +173,11 @@
 			Texture that specifies how much surface emits light at a given point.
 		</member>
+		<member name="fallback_mode" type="int" setter="set_fallback_mode" getter="get_fallback_mode" enum="SpatialMaterial.FallbackMode" default="2">
+			When the project setting [code]rendering/gles3/shaders/async_compile_enabled[/code] is [code]true[/code], this determines how this material must behave in regards to asynchronous shader compilation.
```
```suggestion
			If [member ProjectSettings.rendering/gles3/shaders/async_compile_enabled] is [code]true[/code], this determines how this material must behave in regards to asynchronous shader compilation.
```

RandomShaper replied:
Done!

Calinou commented on doc/classes/SpatialMaterial.xml:
```diff
@@ -173,6 +173,11 @@
 		</member>
+		<member name="fallback_mode" type="int" setter="set_fallback_mode" getter="get_fallback_mode" enum="SpatialMaterial.FallbackMode" default="2">
+			When the project setting [code]rendering/gles3/shaders/async_compile_enabled[/code] is [code]true[/code], this determines how this material must behave in regards to asynchronous shader compilation.
+			This behaves like a hint, which means that, depending on the features this material uses, using the simple fallback may not be possible.
```
```suggestion
			[member fallback_mode] is a hint. This means that depending on the features this material uses, using the simple fallback may not be possible.
```

RandomShaper replied:
Done!

Calinou commented on doc/classes/SpatialMaterial.xml:
```diff
@@ -173,6 +173,11 @@
+		<member name="fallback_mode" type="int" setter="set_fallback_mode" getter="get_fallback_mode" enum="SpatialMaterial.FallbackMode" default="2">
+			When the project setting [code]rendering/gles3/shaders/async_compile_enabled[/code] is [code]true[/code], this determines how this material must behave in regards to asynchronous shader compilation.
+			This behaves like a hint, which means that, depending on the features this material uses, using the simple fallback may not be possible.
+			The default is [constant FALLBACK_MODE_UBERSHADER], which is a good starting point for most projects. Feel free to check the other fallback modes.
```
```suggestion
			The default is [constant FALLBACK_MODE_UBERSHADER], which is a good starting point for most projects.
```

RandomShaper replied:
Done!

Calinou commented on doc/classes/SpatialMaterial.xml:
```diff
@@ -639,5 +644,14 @@
 			Smoothly fades the object out based on the object's distance from the camera using a dither approach. Dithering discards pixels based on a set pattern to smoothly fade without enabling transparency. On certain hardware this can be faster than [constant DISTANCE_FADE_PIXEL_ALPHA].
 		</constant>
+		<constant name="FALLBACK_MODE_NONE" value="0" enum="FallbackMode">
+			This material won't have a fallback. The application will stop to compile the real conditioned shader when it's used for the first time.
```
```suggestion
			This material won't have a fallback. The project will stop to compile the real conditioned shader when it's used for the first time. This can lead to the project temporarily freezing during gameplay.
```

RandomShaper replied:
Done!

Calinou commented on doc/classes/SpatialMaterial.xml:
```diff
@@ -639,5 +644,14 @@
+			Anything with this material applied won't be rendered while this material's shader is being compiled.
+		</constant>
+		<constant name="FALLBACK_MODE_UBERSHADER" value="2" enum="FallbackMode">
+			The real conditioned shader needed on each situation will be sent for background compilation, but in the meantime a very complex shader that adapts to every situation will be used. It will be much slower to render, but will keep the game running without stalling to compile.
```
```suggestion
			The real conditioned shader needed on each situation will be sent for background compilation. In the meantime, a very complex shader that adapts to every situation will be used ("ubershader"). This ubershader is much slower to render, but will keep the game running without stalling to compile. Once shader compilation is done, the ubershader is replaced by the traditional optimized shader.
```

RandomShaper replied:
Done!

Calinou commented on doc/classes/VisualServer.xml:
```diff
@@ -2523,6 +2523,14 @@
+			<argument index="0" name="forbidden" type="bool" />
+			<description>
+				In case asynchronous shader compilation is enabled, this controls whether [code]SHADER_FALLBACK_NO_RENDER[/code] is obeyed.
+				A game may want to enable this before taking a screenshot, to ensure everything is visible regardless the readiness of shaders falling back to [i]no render[/i].
```
```suggestion
				You may want to enable this temporarily before taking a screenshot. This ensures everything is visible regardless the readiness of shaders falling back to [i]no render[/i].
```

RandomShaper replied:
Done!

Calinou commented on doc/classes/VisualServer.xml:
```diff
@@ -2523,6 +2523,14 @@
+			<return type="void" />
+			<argument index="0" name="forbidden" type="bool" />
+			<description>
+				In case asynchronous shader compilation is enabled, this controls whether [code]SHADER_FALLBACK_NO_RENDER[/code] is obeyed.
```
```suggestion
				If asynchronous shader compilation is enabled, this controls whether [member SpatialMaterial.SHADER_FALLBACK_NO_RENDER] is obeyed.
```

RandomShaper replied:
Done!

RandomShaper replied:
Done!

Calinou commented on scene/resources/material.cpp:
```diff
@@ -2136,6 +2163,8 @@ void SpatialMaterial::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::REAL, "distance_fade_min_distance", PROPERTY_HINT_RANGE, "0,4096,0.01"), "set_distance_fade_min_distance", "get_distance_fade_min_distance");
 	ADD_PROPERTY(PropertyInfo(Variant::REAL, "distance_fade_max_distance", PROPERTY_HINT_RANGE, "0,4096,0.01"), "set_distance_fade_max_distance", "get_distance_fade_max_distance");
 
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "fallback_mode", PROPERTY_HINT_ENUM, "Ubershader,None,No Render"), "set_fallback_mode", "get_fallback_mode");
```
Reorder to match the actual enum and improve the hint descriptions:

```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::INT, "fallback_mode", PROPERTY_HINT_ENUM, "None (Synchronous),No Render (Async. Invisible),Ubershader (Async. Visible)"), "set_fallback_mode", "get_fallback_mode");
```

RandomShaper replied:
I've added your descriptions, but kept my ordering (it seems I had already pushed a change of ordering when you had already seen this).

QbieShay commented:
Fallback `none` and `no_render` are a bit confusing, perhaps we can find some alternative terms?

RandomShaper commented:
> Fallback `none` and `no_render` are a bit confusing, perhaps we can find some alternative terms?

Good point. Taking inspiration from @Calinou's suggested changes, I'm thinking that a great rename could happen, like this:

- Fallback mode -> Async mode
- Fallback no render -> Async mode invisible
- Fallback ubershader -> Asyc mode visible slow
- Fallback none -> Async mode disabled

Something along those lines. Thoughts?

QbieShay commented:
@RandomShaper I'm not very convinced. What about:

Shader Compilation
- Hide until ready (fast)
- Simplified fallback (intermediate)
- Synchronous (slow)

Something like this?

tinmanjuggernaut commented:
Adding async isn't more clear. And names like slow, fast, etc aren't descriptive. I want to know what will visually appear on screen. 

Fallback is clear. This will be used while the full shader is being compiled. 

Don't render / invisible is clear. 

The other states aren't descriptive. 

Rather than offer none/no shader, why not default to a simplified albedo shader and give me a color picker, and call it color. 

What is Uber shader? It doesn't mean anything to me. Custom means something. 

I would call it a fallback shader, with choice of: "don't render, color, custom"

QbieShay commented:
@tinmanjuggernaut That was the previous approach, and it adds more complexity and doesn't work as well as an ubershader for most cases.

RandomShaper commented:
> Shader Compilation
> 
> * Hide until ready (fast)
> * Simplified fallback (intermediate)
> * Synchronous (slow)

Those look like more intuitive. However, _Simplified fallback (intermediate)_ doesn't quite fit. The fallback is slower, but no simpler. Also, it'd be great that the first two are somehow understood as asynchronous modes. _Slow until ready_ would meet, but sounds a bit strange, doesn't it?

RandomShaper committed:
Add ProjectSettings::set_hide_from_editor()

RandomShaper commented:
Pushed with documentation that was missing. Naming will be changed when we find good names.

clayjohn commented:
Regarding the naming discussion, is there even a good reason to expose fallback modes? To me it makes the most sense to have a single flag "hide until shader compiled".

I can't see a reason why anyone would want the entire game to freeze up to wait for a shader to compile over using the Ubershader.

RandomShaper commented:
At some point I was asking myself what uses cases would the forced sync support. I couldn't think of any, but I assumed flexibility was better, just in case. By bringing that up you have made me finally believe that it's better just to get rid of it. If one day some user comes across a legit use case, we can always reintroduce it.

If we go that route, would we really want to switch from an enumerated type to a boolean, instead of keeping the enum with two values? The latter, in case of a future addition of new fallback modes, would avoid breaking compatibility. I don't have a strong opinion on this.

tinmanjuggernaut commented:
Oh, the fall backs are blocking? Invisible is ok, but it might reveal hidden elements in a level design. A non-blocking or already compiled universal shader with different opaque colors is better. A complex, blocking fallback shader is unnecessary. If I wanted a blocking shader, I'd just use the stock renderer.  

In addition to async compilation, it would be ideal to have a mechanism to force shader compilation before scene load or on game startup, and a signal on complete. I'm doing this with hacky work arounds described in the other PR, but it would be better if built into the engine.

clayjohn commented:
> If we go that route, would we really want to switch from an enumerated type to a boolean, instead of keeping the enum with two values? The latter, in case of a future addition of new fallback modes, would avoid breaking compatibility. I don't have a strong opinion on this.

This sounds like speculative future-proofing to me. But given tinmanjuggernaut's comments above, is probably a good idea. 

> Oh, the fall backs are blocking? Invisible is ok, but it might reveal hidden elements in a level design. A non-blocking or already compiled universal shader with just opaque colors is better. 

Only "Fallback none" is blocking, the Ubershader approach loads instantly, looks the same as regular materials, but has worse per-frame performance. 

I like the idea of the opaque color fallback (like in RandomShaper's last PR), but I have a feeling it will be unnecessary. In my opinion, we should merge this PR with just the Ubershader and "no render" options and then see if there is a need for something in the middle.

> A complex, blocking fallback shader is unnecessary. If I wanted a blocking shader, I'd just use the stock renderer.

I agree, but oof :P 

> In addition to async compilation, it would be ideal to have a mechanism to force shader compilation before scene load or on game startup, and a signal on complete. I'm doing this with hacky work arounds described in the other PR, but it would be better if built into the engine.

We have discussed precompiling shaders many times and the consensus is that it would just be too hard given hoe Godot is designed.

tinmanjuggernaut commented:
> We have discussed precompiling shaders many times and the consensus is that it would just be too hard given how Godot is designed.

Which was discussed? 
1. Precompiling at dev time and storage? sure. 
2. Precompiling at game start? 
3. Precompiling on scene load before _ready? 
4. Precompiling on demand? 

@tcoxon and I are already doing 3,  4 in gdscript. I traverse the tree, get every material, display it on screen, plus move the camera to a birds eye view, and a 360 rotation to *attempt* to force compile every one, then send a signal. Some of this could definitely be done directly in the engine more cleanly such as identifying every loaded material, and a direct initiation of shader compilation.

RandomShaper commented:
I've not been able to reproduce the glitch –that only happens on a few frames– in 40 runs or so. I give up. Maybe it's a driver bug? My gut instinct is that, in case it's a code issue, it will only happen if async. compilation is enabled. And that's acceptable, given it's disabled by default as a somewhat experimental feature. If some project can find a solid repro case when this is released, I'll try to find it.

Therefore, hereby, I'm considering this ready for merging.

Calinou commented:
Some testing done on a GTX 1080 on Linux:

Enabling async compile helps reduce stuttering when firing a bullet for the first time in the TPS demo.
Artifacts are sometimes visible in the main menu (character's vertices look like they spazz out).

Average FPS on TPS demo with 2560×1440 resolution and maximum settings after starting the level and waiting 10 seconds:

- Async disabled, cache disabled: 45 FPS
- Async **enabled**, cache disabled: 45 FPS
- Async disabled, cache **enabled**: 44 FPS
- Async **enabled**, cache **enabled**: 44 FPS

Performance seems to be about the same overall, although it seems that enabling the cache has a small negative impact on performance. Can anyone else reproduce this?

RandomShaper commented:
> I've not been able to reproduce the glitch –that only happens on a few frames– in 40 runs or so. I give up.

Giving up was not like me. I found the issue. Pushed with a fix plus a bit of additional polishing to the code.

Calinou commented:
I've tested this PR again and it seems that your fix works. I can no longer reproduce the issue locally after several attempts.

RandomShaper commented:
Pushed with the change to the cache path getter on Windows, as discussed in #53580.

akien-mga commented:
I made a test build for all platforms (minus UWP) including export templates:

https://downloads.tuxfamily.org/godotengine/testing/3.4-beta-pr53411-bef883/

It's based on bef883e, so not the very last state, but the only difference is the location of the Windows cache.

owstetra commented:
With this i see not stutter at all with the TPS Demo !
Thank you so much !! you did a great Job @RandomShaper !!
Thank you @akien-mga for the build !!
You guys are awesome !

RandomShaper commented:
It seems Firefox doesn't support/expose `KHR_parallel_shader_compile`. In my case, it gets listed in https://www.geeks3d.com/webgl/ with Chrome and not with Firefox. Weird, since Firefox allegedly has been supporting it for a while: https://caniuse.com/mdn-api_khr_parallel_shader_compile

eon-s commented:
> 
> 
> It seems Firefox doesn't support/expose `KHR_parallel_shader_compile`. In my case, it gets listed in https://www.geeks3d.com/webgl/ with Chrome and not with Firefox. Weird, since Firefox allegedly has been supporting it for a while: https://caniuse.com/mdn-api_khr_parallel_shader_compile

Yes, it is supposed to support it since version 80 but I checked in https://webglreport.com/ even with draft extensions enabled, and is not shown there.

akien-mga commented:
Maybe we don't query the extension the right way? https://developer.mozilla.org/en-US/docs/Web/API/KHR_parallel_shader_compile

tcoxon commented:
@RandomShaper 
> You can enable the force shader fallbacks debug option to see how your game runs once there a no more compiles in progress. How well or badly it runs will be just a consequence of the weight of the ubershaders.

I get 60 FPS with force shader fallbacks on, so I think the dips in the the FPS charts are where the queue is hitting the limit, rather than ubershaders degrading performance.

RandomShaper commented:
Pushed with:
- Fixed wrongly opening GL ES 2 library on mobile. (Good catch, @clayjohn.)
- Fixed issue with normal/tangent attribute in the ubershader.
- Added secondary GL context to Android (so async compile is still possible when true parallel compile is not supported). (@eon-s, if you build yourself, you can test it. My Android Adreno GPU drivers don't expose parallel compile either, so I've tested the new approach with a secondary context. Not super good, but still much better than the stutter.)

clayjohn reviewed (approved):
Looks really good! This is an amazing piece of work. I really like the syntax for choosing between ubershader skip and runtime. 

I tried to carefully review all the conditions in scene.glsl as I am sure there is some combination of flags that will cause an issue, but I didn't see anything. We will need to rely on heavy user testing. 

I suggest we merge this before implementing ubershaders for CanvasItem shaders.

RandomShaper commented:
@Calinou, I went back to a lower size because now only ubershaders are cached. Also, I did a change in how the GLSL code is generated so it's deterministic, which also reduced the storage needs compared to earlier iterations or the PR (formerly, the same shader may be being re-cached on every run, with a different hash). If I recall correctly, the TPS demo in its entirety now takes ~100 MiB of cache with my Nvidia GTX drivers.

Regarding the idea of the enumerated setting, I like it. I'll make the change in an upcoming push.

tinmanjuggernaut commented:
My project w/ 4GB in assets and ~2GB VRAM currently has 122 files and 218mb in the cache dir. As long as it is customizable, it's fine.

tinmanjuggernaut commented:
I've continued to use the binary that Remi made with this PR, and the only problem I have with it is that shader caching isn't enabled in the editor. Can we get that enabled in the next push? Every time I load the engine and my scenes, the editor seizes moving the camera around. Otherwise, I haven't had any problems with the PR itself. I'm using async_compile, and 8 max simultaneous compiles on a GTX1060. 2 seemed sluggish, but I didn't really test it and quickly changed it to 8 and haven't changed it since.

RandomShaper commented:
Pushed with this done:
> Now that synchronous shader compilation + caching is no longer a valid combination for the project setting, we should consider moving the project setting to an enum such as **Shader Compilation Mode**:

That has required modifications to the documentation. Please take a look.

The only pending thing is to debug an issue on Android. @akien-mga, this can be merged right away and I can submit the fix for Android as a separate PR (or, in the worst case, just a patch to disable async on mobile if no true parallel compile is supported).

akien-mga commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1222,6 +1222,39 @@
+		<member name="rendering/gles3/shaders/log_active_async_compiles_count" type="bool" setter="" getter="" default="false">
+			If [code]true[/code], every time an asynchronous shader compilation or an asynchronous shader reconstruction from cache starts or finishes, a line will be logged telling how many of those are happening.
+			If the platform doesn't support parallel shader compile, but only the compile queue via a secondary GL context, what the message will tell is the number of shader compiles currently queued.
+			[b]Note:[/b] This setting is only meaningful if [code]rendering/gles3/shaders/async_compilation_mode[/code] is [b]not[/b] [code]Synchronous[/code].
```
```suggestion
			[b]Note:[/b] This setting is only meaningful if [member rendering/gles3/shaders/shader_compilation_mode] is [b]not[/b] [code]Synchronous[/code].
```
Same for others in this class.

RandomShaper replied:
Coming in next push.

akien-mga commented on doc/classes/SpatialMaterial.xml:
```diff
@@ -87,6 +87,10 @@
 			Specifies the channel of the [member ao_texture] in which the ambient occlusion information is stored. This is useful when you store the information for multiple effects in a single texture. For example if you stored metallic in the red channel, roughness in the blue, and ambient occlusion in the green you could reduce the number of textures you use.
 		</member>
+		<member name="async_mode" type="int" setter="set_async_mode" getter="get_async_mode" enum="SpatialMaterial.AsyncMode" default="0">
+			If [member ProjectSettings.rendering/gles3/shaders/async_compile_enabled] is [code]true[/code], this determines how this material must behave in regards to asynchronous shader compilation.
```
```suggestion
			If [member ProjectSettings.rendering/gles3/shaders/shader_compilation_mode] is [code]true[/code], this determines how this material must behave in regards to asynchronous shader compilation.
```

RandomShaper replied:
Coming in next push.

RandomShaper replied:
Also changed the _is [code]true[/code]_ part, since it's an enum now.

akien-mga commented on platform/android/java/lib/src/org/godotengine/godot/GodotView.java:
```diff
@@ -156,24 +164,49 @@ private void init(XRMode xrMode, boolean translucent, int depth, int stencil) {
 				if (GLUtils.use_32) {
-					setEGLConfigChooser(translucent
+					eglConfigChooser = translucent
 									? new RegularFallbackConfigChooser(8, 8, 8, 8, 24, stencil,
 											  new RegularConfigChooser(8, 8, 8, 8, 16, stencil))
 									: new RegularFallbackConfigChooser(8, 8, 8, 8, 24, stencil,
```
clang-format and ternary operators, fun... now that it's no longer a function call it can be aligned with two tabs as configured, no longer four tabs as some arbitrary approximation of the starting parenthesis.

RandomShaper replied:
Amended!

akien-mga commented:
Thanks! :tada:

tinmanjuggernaut commented:
As of fee83d02cc9b3ea3fcf62737490b1b5a3464ffc5, it looks like this patch is still not enabled in the editor and has no option to enable it, so it lags terribly when opening large scenes.

RandomShaper commented:
Enabling this in the editor involves some design decisions (separate settings for editor and game?, using ubershader is OK since in summer cases it's not 100% equal to the real shader?, maybe it's better to just enable a cache of regular shaders with no async support, or allow async with no ubershader but some placeholder shader?, etc.).

So, while I agree it may be a good thing, it's not trivial. You may want to open a proposal with your choices so a discussion can happen to see what the most wanted behavior is.

tinmanjuggernaut commented:
> So, while I agree it may be a good thing, it's not trivial. You may want to open a proposal with your choices so a discussion can happen to see what the most wanted behavior is.

I have started the proposal to discuss it here:
https://github.com/godotengine/godot-proposals/issues/3594

jitspoe commented:
I know this is kind of old, but I'm still using 3.x on one of my projects and I have a few questions/issues with this.  If I'm understanding this properly: Every shader gets an "ubershader" that accounts for all the possible defines that can show up in the shader (ENABLE_OCTAHEDRAL_COMPRESSION, USE_SKELETON, USE_SHADOW, SHADELESS, etc) at run time.  These are compiled at load time then specific shaders based on the context of the scene are compiled asynchronously and swapped out when done.  This effectively means every shader gets compiled twice, unless "async_hidden" is set, in which case, it seems the ubershader is not used and the mesh is just hidden until the async shader compile is done.

The first time I attempted to enable async, it took an incredibly long time to load the game -- longer than my manual building of the shaders by flashing everything in front of the camera.  I came back and gave it a second shot, but this time I made all the shader async_hidden.  This seems much more ideal.

What I'd like to do is, at the main menu, cycle through all the materials and get them started building asynchronously so there's not a long initial loading screen.  Worst case, a player might jump through the menus really quickly and start a game before all the shaders are build and some stuff would pop in, but that seems like a better tradeoff than a long load every time you open the game.

What I'm unsure of is, if the async limit is 2, does that mean if 3 materials are shown, then hidden, 2 will start building and the 3rd won't build again until it's shown again?  Or does each shader get queued and will eventually build?

Also, why are only the ubershaders cached?  It seems like the nVidia drivers are doing some caching of their own, which is nice.  I can actually run through the initial build of all the materials in like 1 second after my drivers cache, even when async is disabled, but not all drivers handle this that well.

Another thing I've noticed that's strange is that I'll get the async building log for materials that have already been built.  Like I'll switch levels and stuff that was already built will "re-build", but it seems to be instantaneous, so I don't know that it's doing a full build, or maybe the driver has it cached?  Putting breakpoints in the engine, I see that it's getting to the part of the code where it's assembling the text for the shader.  I don't get there when async is disabled and I reload levels.  Seems like this might be a bug where it's re-building text for shaders that doesn't need to be done?

Finally, I ran into some issues with async_hidden where, when some new materials would get built, some old materials would flicker (in my case, they turned grey for 1 frame).  Not all the materials.  Just 1 or 2.  So maybe the ground would flash grey, or the sky, even though those were already built and rendering.
